File: runtime/apps/gui/sdk/gateway/src/types/product.ts
export type ProductConfig = {
  deployment_mode: string;
  signup_enabled: boolean;
  google_oauth_enabled: boolean;
  version: string;
};

export type ProductUser = {
  id: string;
  email: string;
  name: string;
  avatar_url?: string | null;
  language: string;
  timezone: string;
  onboarded_at?: number | null;
};

export type Workspace = {
  id: string;
  name: string;
  slug: string;
  description?: string | null;
  context?: string | null;
  issue_prefix: string;
  avatar?: string | null;
  created_at: number;
  updated_at: number;
};

export type ProductIssueStatus = "backlog" | "todo" | "in_progress" | "review" | "done" | "closed";
export type ProductIssuePriority = "low" | "medium" | "high" | "urgent";

export type TaskRun = {
  id: string;
  issue_id?: string | null;
  agent_id: string;
  runtime_id?: string | null;
  status: string;
  session_id?: string | null;
  title: string;
  created_at: number;
  updated_at: number;
};

export type ProductIssue = {
  id: string;
  workspace_id: string;
  number: number;
  title: string;
  description: string;
  status: ProductIssueStatus;
  priority: ProductIssuePriority;
  position: number;
  assignee_type?: string | null;
  assignee_id?: string | null;
  project_id?: string | null;
  labels: string[];
  session_id?: string | null;
  active_task?: TaskRun | null;
  created_at: number;
  updated_at: number;
};

export type ProductIssueInput = {
  title?: string;
  description?: string;
  status?: ProductIssueStatus;
  priority?: ProductIssuePriority;
  assignee_type?: string | null;
  assignee_id?: string | null;
  project_id?: string | null;
  labels?: string[];
  session_id?: string | null;
};

export type ProductProject = {
  id: string;
  workspace_id: string;
  title: string;
  description: string;
  status: string;
  priority: string;
  lead_type?: string | null;
  lead_id?: string | null;
  created_at: number;
  updated_at: number;
};

export type ProductAgent = {
  id: string;
  workspace_id: string;
  name: string;
  description: string;
  provider: string;
  model: string;
  runtime_id?: string | null;
  status: string;
  visibility: string;
  thinking_level?: string | null;
  run_count_7d: number;
  run_count_30d: number;
};

export type RuntimeDevice = {
  id: string;
  workspace_id: string;
  provider: string;
  name: string;
  runtime_mode: string;
  visibility: string;
  status: string;
  last_seen_at: number;
  cli_version?: string | null;
  launched_by?: string | null;
};

export type InboxItem = {
  id: string;
  workspace_id: string;
  type: string;
  severity: string;
  title: string;
  issue_id?: string | null;
  read_at?: number | null;
  archived_at?: number | null;
  created_at: number;
};

export type UsagePoint = {
  date: string;
  tasks: number;
  tokens: number;
  cost: number;
};

export type UsageByAgent = {
  agent_id: string;
  tasks: number;
  tokens: number;
  cost: number;
};

export type AgentRuntimeUsage = {
  agent_id: string;
  runtime_id?: string | null;
  active_tasks: number;
  status: string;
};

File: runtime/crates/gateway/src/contracts/product.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize)]
pub struct PublicConfig {
    pub deployment_mode: String,
    pub signup_enabled: bool,
    pub google_oauth_enabled: bool,
    pub version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProductUser {
    pub id: String,
    pub email: String,
    pub name: String,
    pub avatar_url: Option<String>,
    pub language: String,
    pub timezone: String,
    pub onboarded_at: Option<i64>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UserPatch {
    pub name: Option<String>,
    pub language: Option<String>,
    pub timezone: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Workspace {
    pub id: String,
    pub name: String,
    pub slug: String,
    pub description: Option<String>,
    pub context: Option<String>,
    pub issue_prefix: String,
    pub avatar: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum IssueStatus {
    Backlog,
    Todo,
    InProgress,
    Review,
    Done,
    Closed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum IssuePriority {
    Low,
    Medium,
    High,
    Urgent,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Issue {
    pub id: String,
    pub workspace_id: String,
    pub number: u32,
    pub title: String,
    pub description: String,
    pub status: IssueStatus,
    pub priority: IssuePriority,
    pub position: i64,
    pub assignee_type: Option<String>,
    pub assignee_id: Option<String>,
    pub project_id: Option<String>,
    pub labels: Vec<String>,
    pub session_id: Option<String>,
    pub active_task: Option<TaskRun>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IssueInput {
    pub title: Option<String>,
    pub description: Option<String>,
    pub status: Option<IssueStatus>,
    pub priority: Option<IssuePriority>,
    pub assignee_type: Option<String>,
    pub assignee_id: Option<String>,
    pub project_id: Option<String>,
    pub labels: Option<Vec<String>>,
    pub session_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IssueQuery {
    pub workspace_id: Option<String>,
    pub workspace_slug: Option<String>,
    pub status: Option<IssueStatus>,
    pub search: Option<String>,
    pub project_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProductProject {
    pub id: String,
    pub workspace_id: String,
    pub title: String,
    pub description: String,
    pub status: String,
    pub priority: String,
    pub lead_type: Option<String>,
    pub lead_id: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProductAgent {
    pub id: String,
    pub workspace_id: String,
    pub name: String,
    pub description: String,
    pub provider: String,
    pub model: String,
    pub runtime_id: Option<String>,
    pub status: String,
    pub visibility: String,
    pub thinking_level: Option<String>,
    pub run_count_7d: u32,
    pub run_count_30d: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskRun {
    pub id: String,
    pub issue_id: Option<String>,
    pub agent_id: String,
    pub runtime_id: Option<String>,
    pub status: String,
    pub session_id: Option<String>,
    pub title: String,
    pub created_at: i64,
    pub updated_at: i64,
}
